File: eCommerce/ClientApp/src/app/shared/classes/product.ts
import { Brand } from "./brands";
import { Bundle } from "./bundle";
import { Category, SubCategory } from "./categories";
import { Operations } from "./enums/operations";
import { Trademarks } from "./enums/trademarks";
import { Review } from "./review";

// Products
export interface Product {
  id?: number;
  title?: string;
  description?: string;
  about?:string;
  type?: string;
  brand?: string;
  collection?: any[];
  category?: string;
  subCategory?: string;
  price?: number;
  sale?: boolean;
  discountRate?: number;
  stock?: number;
  new?: boolean;
  quantity?: number;
  customerPreferenceQuantity?: number;
  tags?: any[];
  variants?: Variants[];
  images?: Images[];
  bundle?: Bundle;
}
export interface SpecialProduct {
  id?: number;
  title?: string;
  quantity?: number;
  price?: number;
  discountRate?: number;

  rating?: any;
  ratingVotes?: number;

  marketStatus?: MarketStatus;
  images?: Images[];
}


export interface Products {
  id?: number;
  title?: string;

  description?: string;
  usage?: string;
  composition?: string;
  characteristic?: string;
  details?: string;

  videoUrl?: string;
  quantity?: number;
  customerPreferenceQuantity?: number;
  sold?: number;
  price?: number;
  category?: Category;
  categoryTitle?: string;
  categoryId?: number;
  subCategoryTitle?: string;
  subCategoryId?: number;
  brandId?: number;
  brandName?: string;
  brand?: Brand;
  discountRate?: number;
  userId?: string;
  isFreeShipping?: boolean;
  isReturnRequestAllowed?: boolean;
  views?: number;
  rating?: any;
  ratingVotes?: number;
  positiveRating?: any;
  premiumPackage?: number;
  marketStatus?: MarketStatus;
  trademark?: Trademarks;
  approveType?: number;
  itemType?: number;
  status?: number;
  gender?: number;
  createdOn?: Date;
  expiredOn?: Date;
  src?: string;

  variants?: Variants[];
  images?: Images[];

  categories?: Category[];
  subCategory?: SubCategory[];
  reviews?: Review[];
  bundle?: Bundle;

  activeColor?: any;
  activeSize?: any;

  isLoading?: boolean;
}


export class Variant {
  title?: string;
  value?: string;
}
export class SelectedFilterArray {
  title?: string;
  values?: string[];
}
export class FilterVariants {
  id?: number;
  title?: string;
  icon?: string;
  variantItems?: VariantItem[];
}
export class Variants {
  productId?: number;
  id?: number;
  size?: string;
  color?: string;
  image?: string;
  value?: string;
  title?: string;
  isSelected?: boolean;
  icon?: string;
  variantItems?: VariantItem[];
}
export class VariantItem {
  id?: number;
  //variantId?: number;
  //variant?: Variants;
  value?: string;
  image?: string;
  isSelected?: boolean;
}

export class Images {
  constructor(operation: Operations){
    this.Operation = operation
  }

  productId?: number;
  id?: number;
  isExternal?: boolean;
  imageType?:ImageType;
  src?: string;
  variantId?: number;

  Operation?: Operations;
}
export class ProductIds {
  quantity?: number;
  id?: number;
}

export class LocalStorageProduct {
  key?: string;
  value?: any;
}
export class LocalStorageImages {
  productId?: number;
  src?: string;
  alt?: string;
}
export class LocalStorageVariants {
  productId?: number;
  color?: string;
  size?: string;
  imageId?: string;
}

export enum MarketStatus {
  None = 0,
  New = 1,
  Sale = 2,
  PreOrder = 3,
  Sold = 4,
  Low = 5
}

export enum PremiumPackage {
  None = 0,
  Bronze = 1,
  Silver = 2,
  Gold = 3,
}
export enum PurchaseStatus {
  None = 0,
  Quantity = 1,
  Processing = 2,
  Finish = 3,
}

export enum ImageType {
  Thumb = 0,
  Main = 1,
  About = 2,
  Usage = 3,
  Composition = 4,
  Characteristic = 5
}

export const ImageType_LabelMapping: Record<ImageType, string> = {
  [ImageType.Thumb]: "Thumb",
  [ImageType.Main]: "Main",
  [ImageType.About]: "About",
  [ImageType.Usage]: "Usage",
  [ImageType.Composition]: "Composition",
  [ImageType.Characteristic]: "Characteristic",
};

File: Entities/Models/Product.cs
using Ardalis.GuardClauses;
using Entities.Enums;
using System.ComponentModel.DataAnnotations.Schema;

namespace Entities.Models
{
    public class Product
    { 
        public int Id { get; set; }
        public string Title { get; set; }
        public string ShortName { get; set; }
        public string Details { get; set; }
        public string Usage { get; set; }
        public string Characteristic { get; set; }
        public string Composition { get; set; }
        public string Description { get; set; }
        public string VideoUrl { get; set; }

        public int Quantity { get; set; }
        public double Price { get; set; }
        public double DiscountRate { get; set; }

        public bool IsFreeShipping { get; set; }
        public bool IsReturnRequestAllowed { get; set; }

        public int? Views { get; set; } = 0;
        public int? Sold { get; set; } = 0;

        public double Rating { get; set; } = 0.0;
        public double PositiveRating { get; set; } = 0.0;

        public int RatingVotes { get; set; } = 0;
        public int VotedUsers { get; set; } = 0;

        public PremiumPackage PremiumPackage { get; set; }
        public ApproveType ApproveType { get; set; }
        public ItemType ItemType { get; set; }
        public Status Status { get; set; }
        public Gender Gender { get; set; }
        public MarketStatus MarketStatus { get; set; }
        public Trademarks Trademark { get; set; }

        public string CreatedOn { get; set; }
        public string ExpiredOn { get; set; }

        [NotMapped]
        public string StatementType { get; set; }

        public void Update(ProductInput input)
        {
            Id = input.Id;

            Guard.Against.NullOrEmpty(input.Title, nameof(input.Title));
            Title = input.Title;

            Guard.Against.NullOrEmpty(input.ShortName, nameof(input.ShortName));
            ShortName = input.ShortName;

            Guard.Against.NullOrEmpty(input.Details, nameof(input.Details));
            Details = input.Details;

            Guard.Against.NullOrEmpty(input.Description, nameof(input.Description));
            Description = input.Description;

            Characteristic = input.Characteristic;
            Composition = input.Composition;
            Usage = input.Usage;

            VideoUrl = input.VideoUrl;
            IsReturnRequestAllowed = input.IsReturnRequestAllowed;

            ApproveType = ApproveType.Success;

            ItemType = (ItemType)Enum.Parse(typeof(ItemType), input.ItemType.ToString());
            Status = (Status)Enum.Parse(typeof(Status), input.Status.ToString());
            Gender = (Gender)Enum.Parse(typeof(Gender), input.Gender.ToString());
            MarketStatus = (MarketStatus)Enum.Parse(typeof(MarketStatus), input.MarketStatus.ToString());
            Trademark = (Trademarks)Enum.Parse(typeof(Trademarks), input.Trademark.ToString());
            PremiumPackage = (PremiumPackage)Enum.Parse(typeof(PremiumPackage), input.PremiumPackage.ToString());

            Guard.Against.NegativeOrZero(input.Quantity, nameof(input.Quantity));
            Quantity = input.Quantity;

            Guard.Against.NegativeOrZero(input.Price, nameof(input.Price));
            Price = input.Price;

            DiscountRate = input.DiscountRate;

            CreatedOn = DateTime.Now.ToString("dd/MM/yyyy");
            ExpiredOn = DateTime.Now.AddDays(90).ToString("dd/MM/yyyy");

        }

    }
}
